blockprod/src/detail/mod.rs

```diff
@@ -141,14 +149,102 @@ impl BlockProduction {
                         })
                     };
 
+                    let block_height = best_block_index.block_height().next_height();
+                    let sealed_epoch_index = chain_config.sealed_epoch_index(&block_height);
+
+                    let sealed_epoch_randomness = sealed_epoch_index
+                        .map(|index| this.get_epoch_data(index))
+                        .transpose()
+                        .map_err(|_| {
+                            ConsensusPoSError::PropertyQueryError(
+                                PropertyQueryError::EpochDataNotFound(block_height),
+                            )
+                        })?
+                        .flatten()
+                        // TODO: no epoch_data means either that no
+                        // epoch has been created yet, or that the
+                        // data is actually missing
+                        .or_else(|| Some(EpochData::new(PoSRandomness::at_genesis(&chain_config))));
+
                     let consensus_data = consensus::generate_consensus_data(
                         &chain_config,
                         &best_block_index,
+                        sealed_epoch_randomness.clone(),
+                        input_data.clone(),
```

Comment:
can this clone be avoided if calculate `finalize_block_data` before?

Reply:
I tried moving things around, but unfortunately it won't work because copies of both `sealed_epoch_randomnness` and `input_data` are needed in more than one place (unless I'm mistaken).